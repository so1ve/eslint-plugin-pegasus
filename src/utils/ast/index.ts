export * from "./call-or-new-expression";
export * from "./is-member-expression";
export * from "./is-method-call";
