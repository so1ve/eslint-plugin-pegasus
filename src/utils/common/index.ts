export * from "./boolean";
export * from "./is-logical-expression";
export * from "./is-node-matches";
export * from "./is-node-value-not-function";
export * from "./is-number";
export * from "./parentheses";
export * from "./rule";
