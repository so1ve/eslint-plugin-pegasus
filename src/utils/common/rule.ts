import type { TSESLint } from "@typescript-eslint/utils";

const isIterable = (object: any): object is Iterable<any> =>
	typeof object?.[Symbol.iterator] === "function";

class FixAbortError extends Error {}
const fixOptions = {
	abort() {
		throw new FixAbortError("Fix aborted.");
	},
};

export type AbortFunction = () => void;

type WrappedFixFunction = (
	fixer: TSESLint.RuleFixer,
	additional: { abort: AbortFunction },
) =>
	| IterableIterator<TSESLint.RuleFix>
	| TSESLint.RuleFix
	| readonly TSESLint.RuleFix[]
	| null;

export const wrapFixFunction =
	(fix: WrappedFixFunction): TSESLint.ReportFixFunction =>
	(fixer) => {
		const result = fix(fixer, fixOptions);

		if (isIterable(result)) {
			try {
				return [...result];
			} catch (error) {
				if (error instanceof FixAbortError) {
					return [];
				}

				throw error;
			}
		}

		return result;
	};
