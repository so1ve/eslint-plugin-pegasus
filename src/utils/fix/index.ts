export * from "./remove-member-expression-property";
export * from "./remove-method-call";
