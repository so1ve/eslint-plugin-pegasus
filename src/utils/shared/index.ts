export * from "./simple-array-search-rule";
