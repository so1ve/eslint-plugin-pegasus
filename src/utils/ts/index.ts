import typeutils from "@typescript-eslint/type-utils";
import type {
	ParserServicesWithTypeInformation,
	TSESTree,
} from "@typescript-eslint/typescript-estree";
import * as tsutils from "ts-api-utils";
import ts from "typescript";

export class TypeHelper {
	typeChecker: ts.TypeChecker;

	constructor(public services: ParserServicesWithTypeInformation) {
		this.typeChecker = services.program.getTypeChecker();
	}

	private getType(node: TSESTree.Node) {
		return typeutils.getConstrainedTypeAtLocation(this.services, node);
	}

	isTypeArrayTypeOrUnionOfArrayTypes(node: TSESTree.Node) {
		const nodeType = this.getType(node);

		return typeutils.isTypeArrayTypeOrUnionOfArrayTypes(
			nodeType,
			this.typeChecker,
		);
	}

	isNumber(node: TSESTree.Node) {
		const nodeType = this.getType(node);

		return tsutils.isTypeFlagSet(nodeType, ts.TypeFlags.Number);
	}

	isAny(node: TSESTree.Node) {
		const nodeType = this.getType(node);

		return typeutils.isTypeAnyType(nodeType);
	}
}
