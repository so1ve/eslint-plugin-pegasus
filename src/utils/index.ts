import { ESLintUtils } from "@typescript-eslint/utils";

export const createRule = ESLintUtils.RuleCreator(
	(ruleName) =>
		`https://github.com/so1ve/eslint-plugin-pegasus/blob/main/docs/rules/${ruleName}.md`,
);
