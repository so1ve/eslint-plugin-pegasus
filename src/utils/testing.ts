import { resolve } from "node:path";

import type { TSESLint } from "@typescript-eslint/utils";
import { expect } from "vitest";

import type { InvalidTestCase } from "../../vendor/rule-tester/src";
import { RuleTester } from "../../vendor/rule-tester/src/RuleTester";

const fixturesDir = resolve(process.cwd(), "test", "fixtures");

export function getRuleTester() {
	const ruleTester = new RuleTester({
		parser: require.resolve("@typescript-eslint/parser"),
		parserOptions: {
			ecmaVersion: 2018,
			tsconfigRootDir: fixturesDir,
			project: "./tsconfig.json",
		},
	});

	return ruleTester;
}

export function testRule<
	MessageIds extends string,
	Options extends readonly unknown[] = [],
	RuleListener extends TSESLint.RuleListener = TSESLint.RuleListener,
>(
	ruleName: string,
	rule: TSESLint.RuleModule<MessageIds, Options, RuleListener>,
	{
		valid,
		invalid,
	}: {
		valid: string[];
		invalid: InvalidTestCase<MessageIds, Options>[];
	},
) {
	const ruleTester = getRuleTester();

	ruleTester.run(ruleName, rule, {
		valid,
		invalid,
	});
}

export function testRuleSnapshot<
	MessageIds extends string,
	Options extends readonly unknown[] = [],
	RuleListener extends TSESLint.RuleListener = TSESLint.RuleListener,
>(
	ruleName: string,
	rule: TSESLint.RuleModule<MessageIds, Options, RuleListener>,
	{
		valid = [],
		invalid = [],
	}: {
		valid?: string[];
		invalid?: string[];
	},
) {
	const ruleTester = getRuleTester();

	ruleTester.run(ruleName, rule, {
		valid,
		invalid: invalid.map((i) => ({
			code: i,
			errors: null,
			onOutput: (output: string) => {
				expect(output).toMatchSnapshot();
			},
		})),
	});
}
