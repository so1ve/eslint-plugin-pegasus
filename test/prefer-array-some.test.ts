import dedent from "dedent";

import type { MessageIds, Options } from "../src/rules/prefer-array-some";
import rule, { RULE_NAME } from "../src/rules/prefer-array-some";
import { testRule, testRuleSnapshot } from "../src/utils/testing";
import type { InvalidTestCase } from "../vendor/rule-tester/src";

const invalidCase = ({
	code,
	suggestionOutput,
	method,
}: {
	code: string;
	suggestionOutput: string;
	method: string;
}): InvalidTestCase<MessageIds, Options> => ({
	code,
	errors: [
		{
			messageId: "some",
			data: { method },
			suggestions: [
				{
					messageId: "someSuggestion",
					data: { method },
					output: suggestionOutput,
				},
			],
		},
	],
});

describe("prefer-array-some", () => {
	testRule(RULE_NAME, rule, {
		valid: [
			// Not `boolean`
			"const bar = foo.find(fn)",
			"const bar = foo.find(fn) || baz",
			"if (foo.find(fn) ?? bar) {}",

			// Not matched `CallExpression`
			...[
				// Not `CallExpression`
				"new foo.find(fn)",
				// Not `MemberExpression`
				"find(fn)",
				// `callee.property` is not a `Identifier`
				'foo["find"](fn)',
				'foo["fi" + "nd"](fn)',
				"foo[`find`](fn)",
				// Computed
				"foo[find](fn)",
				// Not `.find`
				"foo.notFind(fn)",
				// More or less argument(s)
				"foo.find()",
				"foo.find(fn, thisArgument, extraArgument)",
				"foo.find(...argumentsArray)",
			].flatMap((code) => [
				`if (${code}) {}`,
				`if (${code.replace("find", "findLast")}) {}`,
			]),
		],
		invalid: [
			...[
				"const bar = !foo.find(fn)",
				"const bar = Boolean(foo.find(fn))",
				"if (foo.find(fn)) {}",
				"const bar = foo.find(fn) ? 1 : 2",
				"while (foo.find(fn)) foo.shift();",
				"do {foo.shift();} while (foo.find(fn));",
				"for (; foo.find(fn); ) foo.shift();",
			].flatMap((code) => {
				code = `const foo = [];${code}`;

				return [
					invalidCase({
						code,
						suggestionOutput: code.replace("find", "some"),
						method: "find",
					}),
					invalidCase({
						code: code.replace("find", "findLast"),
						suggestionOutput: code.replace("find", "some"),
						method: "findLast",
					}),
				];
			}),
			// Comments
			invalidCase({
				code: "const foo = []; console.log(foo /* comment 1 */ . /* comment 2 */ find /* comment 3 */ (fn) ? a : b)",
				suggestionOutput:
					"const foo = []; console.log(foo /* comment 1 */ . /* comment 2 */ some /* comment 3 */ (fn) ? a : b)",
				method: "find",
			}),
		],
	});

	testRuleSnapshot(RULE_NAME, rule, {
		valid: [],
		invalid: [
			'const array = []; if (array.find(element => element === "ðŸ¦„")) {}',
			'const array = []; const foo = array.find(element => element === "ðŸ¦„") ? bar : baz;',
		],
	});

	// - `.filter(â€¦).length > 0`
	// - `.filter(â€¦).length !== 0`
	testRuleSnapshot(RULE_NAME, rule, {
		valid: [
			// `> 0`
			"array.filter(fn).length > 0.",
			"array.filter(fn).length > .0",
			"array.filter(fn).length > 0.0",
			"array.filter(fn).length > 0x00",
			"array.filter(fn).length < 0",
			"array.filter(fn).length >= 0",
			"0 > array.filter(fn).length",

			// `!== 0`
			"array.filter(fn).length !== 0.",
			"array.filter(fn).length !== .0",
			"array.filter(fn).length !== 0.0",
			"array.filter(fn).length !== 0x00",
			"array.filter(fn).length != 0",
			"array.filter(fn).length === 0",
			"array.filter(fn).length == 0",
			"array.filter(fn).length = 0",
			"0 !== array.filter(fn).length",

			// `>= 1`
			"array.filter(fn).length >= 1",
			"array.filter(fn).length >= 1.",
			"array.filter(fn).length >= 1.0",
			"array.filter(fn).length >= 0x1",
			"array.filter(fn).length > 1",
			"array.filter(fn).length < 1",
			"array.filter(fn).length = 1",
			"array.filter(fn).length += 1",
			"1 >= array.filter(fn).length",

			// `.length`
			"array.filter(fn)?.length > 0",
			"array.filter(fn)[length] > 0",
			"array.filter(fn).notLength > 0",
			"array.filter(fn).length() > 0",
			"+array.filter(fn).length >= 1",

			// `.filter`
			"array.filter?.(fn).length > 0",
			"array?.filter(fn).length > 0",
			"array.notFilter(fn).length > 0",
			"array.filter.length > 0",

			// `jQuery#filter`
			'$element.filter(":visible").length > 0',
		],
		invalid: [
			"array.filter(fn).length > 0",
			"array.filter(fn).length !== 0",
			dedent`
			if (
				((
					((
						((
							((
								array
							))
								.filter(what_ever_here)
						))
							.length
					))
					>
					(( 0 ))
				))
			);
		`,
		],
	});

	// Compare with `undefined`
	testRuleSnapshot(RULE_NAME, rule, {
		valid: [
			"foo.find(fn) == 0",
			'foo.find(fn) != ""',
			"foo.find(fn) === null",
			'foo.find(fn) !== "null"',
			"foo.find(fn) >= undefined",
			"foo.find(fn) instanceof undefined",
			// We are not checking this right now
			'typeof foo.find(fn) === "undefined"',
		],
		invalid: [
			"foo.find(fn) == null",
			"foo.find(fn) == undefined",
			"foo.find(fn) === undefined",
			"foo.find(fn) != null",
			"foo.find(fn) != undefined",
			"foo.find(fn) !== undefined",
			'a = (( ((foo.find(fn))) == ((null)) )) ? "no" : "yes";',
		].map((code) => `const foo = []; ${code}`),
	});
});
