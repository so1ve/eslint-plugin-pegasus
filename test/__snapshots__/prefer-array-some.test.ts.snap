// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prefer-array-some > prefer-array-some > invalid > array.filter(fn).length !== 0 1`] = `"array.some(fn)"`;

exports[`prefer-array-some > prefer-array-some > invalid > array.filter(fn).length > 0 1`] = `"array.some(fn)"`;

exports[`prefer-array-some > prefer-array-some > invalid > const array = []; const foo = array.find(element => element === "ğŸ¦„") ? bar : baz; 1`] = `"const array = []; const foo = array.find(element => element === \\"ğŸ¦„\\") ? bar : baz;"`;

exports[`prefer-array-some > prefer-array-some > invalid > const array = []; if (array.find(element => element === "ğŸ¦„")) {} 1`] = `"const array = []; if (array.find(element => element === \\"ğŸ¦„\\")) {}"`;

exports[`prefer-array-some > prefer-array-some > invalid > const foo = []; a = (( ((foo.find(fn))) == ((null)) )) ? "no" : "yes"; 1`] = `"const foo = []; a = (( ((foo.find(fn))) == ((null)) )) ? \\"no\\" : \\"yes\\";"`;

exports[`prefer-array-some > prefer-array-some > invalid > const foo = []; foo.find(fn) != null 1`] = `"const foo = []; foo.find(fn) != null"`;

exports[`prefer-array-some > prefer-array-some > invalid > const foo = []; foo.find(fn) != undefined 1`] = `"const foo = []; foo.find(fn) != undefined"`;

exports[`prefer-array-some > prefer-array-some > invalid > const foo = []; foo.find(fn) !== undefined 1`] = `"const foo = []; foo.find(fn) !== undefined"`;

exports[`prefer-array-some > prefer-array-some > invalid > const foo = []; foo.find(fn) == null 1`] = `"const foo = []; foo.find(fn) == null"`;

exports[`prefer-array-some > prefer-array-some > invalid > const foo = []; foo.find(fn) == undefined 1`] = `"const foo = []; foo.find(fn) == undefined"`;

exports[`prefer-array-some > prefer-array-some > invalid > const foo = []; foo.find(fn) === undefined 1`] = `"const foo = []; foo.find(fn) === undefined"`;

exports[`prefer-array-some > prefer-array-some > invalid > if (
\\u0009((
\\u0009\\u0009((
\\u0009\\u0009\\u0009((
\\u0009\\u0009\\u0009\\u0009((
\\u0009\\u0009\\u0009\\u0009\\u0009array
\\u0009\\u0009\\u0009\\u0009))
\\u0009\\u0009\\u0009\\u0009\\u0009.filter(what_ever_here)
\\u0009\\u0009\\u0009))
\\u0009\\u0009\\u0009\\u0009.length
\\u0009\\u0009))
\\u0009\\u0009>
\\u0009\\u0009(( 0 ))
\\u0009))
); 1`] = `
"if (
	((
		((
			((
				((
					array
				))
					.some(what_ever_here)
			))
		))
	))
);"
`;
